sfe/overridesimporter.go

```diff
@@ -0,0 +1,302 @@
+	ra      rapb.RegistrationAuthorityClient
+
+	clk clock.Clock
+	log blog.Logger
+}
+
+// NewOverridesImporter creates a new OverridesImporter that will process
+// tickets in the given mode at the given interval. An error is returned if the
+// interval is left unspecified.
+func NewOverridesImporter(mode ProcessMode, interval time.Duration, client *zendesk.Client, sa rapb.RegistrationAuthorityClient, clk clock.Clock, log blog.Logger) (*OverridesImporter, error) {
+	if interval <= 0 {
+		return nil, fmt.Errorf("interval cannot be 0")
+	}
+	return &OverridesImporter{
+		mode:     mode,
+		interval: interval,
+		zendesk:  client,
+		ra:       sa,
+		clk:      clk,
+		log:      log,
+	}, nil
+}
+
+// Start begins the periodic import of approved override requests from Zendesk.
+// This method blocks until the provided context is cancelled.
+func (im *OverridesImporter) Start(ctx context.Context) {
+	ticker := time.NewTicker(im.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			im.tick(ctx)
+		}
+	}
+}
+
+// tick performs a single import pass, serially processing all tickets in the
+// configured mode that have been marked "open" and "approved".
+func (im *OverridesImporter) tick(ctx context.Context) {
+	tickets, err := im.zendesk.FindTickets(map[string]string{ReviewStatusFieldName: reviewStatusApproved}, "open")
+	if err != nil {
+		im.log.Errf("while searching zendesk for solved and approved tickets: %s", err)
+		return
+	}
+
+	processed := 0
+	failures := 0
+	for id, fields := range tickets {
+		switch im.mode {
+		case processEven:
+			if id%2 != 0 {
+				continue
+			}
+		case processOdd:
+			if id%2 == 0 {
+				continue
+			}
+		}
+
+		err = im.processTicket(ctx, id, fields)
+		if err != nil {
+			im.log.Errf("while processing ticket %d: %s", id, err)
+			failures++
+			continue
+		}
+		processed++
+	}
+	im.log.Infof("overrides importer processed %d tickets with %d failures", processed, failures)
+}
+
+func accountURIToID(s string) (int64, error) {
+	u, err := url.Parse(s)
+	if err != nil {
+		return 0, err
+	}
+	parts := strings.Split(strings.Trim(u.Path, "/"), "/")
+	if len(parts) < 3 {
+		return 0, fmt.Errorf("unexpected path %q", u.Path)
+	}
+	id, convErr := strconv.ParseInt(parts[len(parts)-1], 10, 64)
+	if convErr != nil {
+		return 0, convErr
+	}
+	return id, nil
+}
+
+// transitionToPendingWithComment sets the status of the given ticket to
+// "pending" and adds a private comment with the given cause. If updating the
+// ticket fails, the error is logged.
+func (im *OverridesImporter) transitionToPendingWithComment(ticketID int64, cause string) {
+	privateBody := fmt.Sprintf(
+		"A failure occurred while importing this override:\n\n%s\n\n"+
+			"This ticket's status has been set to pending.\n\n"+
+			"Once the error has been corrected, change the status back to \"solved\" to retry.\n",
+		cause,
+	)
+	err := im.zendesk.UpdateTicketStatus(ticketID, "pending", privateBody, false)
```

Comment:
As a possible followup: I'd love to split `UpdateTicketStatus` into `UpdateTicketPublic` and `UpdateTicketPrivate`, since boolean fields can make it easy to make copy-paste mistakes. Definitely not required as part of this PR.

Reply:
I like that idea. I had to write tests to make sure that we got the intended result and it should keyed me in to the idea that these two should be a little more separate.